src-input/duk_bi_encoding.c

```diff
@@ -0,0 +1,348 @@
+			dec_ctx->lower = 0x80;
+			dec_ctx->upper = 0xbf;
+			dec_ctx->codepoint = (dec_ctx->codepoint << 6) | (byte & 0x3f);
+			if (++dec_ctx->seen < dec_ctx->needed) {
+				/* need more bytes */
+				return DUK__U8_CONTINUE;
+			} else {
+				/* got a codepoint */
+				*cp = dec_ctx->codepoint;
+				dec_ctx->codepoint = 0x0000L;
+				dec_ctx->needed = 0;
+				dec_ctx->seen = 0;
+				return DUK__U8_CODEPOINT;
+			}
+		} else {
+			/* We've just seen an invalid UTF-8 continuation byte.  This might
+			 * be the initial byte of the next character; if we return a plain
+			 * error status and the decoder is in replacement mode, the character
+			 * will be masked.  We still need to alert the caller to the error
+			 * though.
+			 */
+			dec_ctx->codepoint = 0x0000L;
+			dec_ctx->needed = 0;
+			dec_ctx->seen = 0;
+			dec_ctx->lower = 0x80;
+			dec_ctx->upper = 0xbf;
+			return DUK__U8_RETRY;
+		}
+	}
+}
+
+DUK_LOCAL void duk__utf8_encode_char(void *udata, duk_codepoint_t codepoint) {
+	duk__encode_context *enc_ctx;
+
+	enc_ctx = (duk__encode_context *) udata;
+	if (codepoint > 0x10ffffL) {
+		/* not a valid UTF-8 codepoint */
+		enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
+	} else if (codepoint >= 0xd800L && codepoint <= 0xdbffL) {
+		/* high surrogate */
+		if (enc_ctx->lead != 0x0000L) {
+			/* consecutive high surrogates, consider first one unpaired */
+			enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
+		}
+		enc_ctx->lead = codepoint;
+	} else if (codepoint >= 0xdc00L && codepoint <= 0xdfffL) {
+		/* low surrogate */
+		if (enc_ctx->lead != 0x0000L) {
+			codepoint = 0x010000L | ((enc_ctx->lead - 0xd800L) << 10) | (codepoint - 0xdc00L);
+			enc_ctx->out += duk_unicode_encode_xutf8(codepoint, enc_ctx->out);
+			enc_ctx->lead = 0x0000L;
+		} else {
+			/* unpaired low surrogate */
+			enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
+		}
+	} else {
+		if (enc_ctx->lead != 0x0000L) {
+			/* unpaired high surrogate */
+			enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
+			enc_ctx->lead = 0x0000L;
+		}
+		enc_ctx->out += duk_unicode_encode_xutf8(codepoint, enc_ctx->out);
+	}
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textencoder_constructor(duk_context *ctx) {
+	/*
+	 *  TextEncoder currently requires no persistent state, so this constructor
+	 *  does nothing on purpose.
+	 */
+
+	if (!duk_is_constructor_call(ctx)) {
+        duk_error(ctx, DUK_ERR_TYPE_ERROR, "constructor must use 'new'");
+    }
+	return 0;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textencoder_prototype_encoding_getter(duk_context *ctx) {
+	duk_push_string(ctx, "utf-8");
+	return 1;
+}
+
+DUK_INTERNAL duk_ret_t duk_bi_textencoder_prototype_encode(duk_context *ctx) {
+	duk__encode_context enc_ctx;
+	duk_size_t len;
+	duk_uint8_t *output;
+
+	if (duk_get_top(ctx) >= 1) {
+		duk_dup(ctx, 0);
```

Comment:
Firefox doesn't seem to distinguish between no argument and undefined btw:

```
new TextEncoder().encode()
Uint8Array [  ]
new TextEncoder().encode(undefined)
Uint8Array [  ]
```

If this is against the WHATWG specification, maybe a one sentence comment why the varargs approach is correct. Current behavior (as you of course know :) differs from Firefox:

```
duk> new TextEncoder().encode()
= ||
duk> new TextEncoder().encode(undefined)
= |756e646566696e6564|
```

For `null` behavior matches.